docs: add READMEs for Slurm integration interface implementations (#239)
charmed-hpc/slurm-charms: internal/slurmrestd-interface/README.md

# `charmed-slurm-slurmrestd-interface`

## Usage

This package provides the integration interface implementation for the `slurmrestd` interface. It enables
`slurmrestd` (Slurm REST API daemon) applications to receive controller data from `slurmctld`,

> including authentication secrets and the Slurm configuration.

Reviewer: ```suggestion
This package provides the integration interface implementation for the `slurmrestd` interface. It enables `slurmrestd` (Slurm REST API daemon) applications to exchange data with `slurmctld`, including authentication secrets and the Slurm configuration.
```

To install, add `charmed-slurm-slurmrestd-interface` to your Python dependencies.
Then in your Python code, import as:

```python
from charmed_slurm_slurmrestd_interface import (
    SlurmrestdProvider,
    SlurmrestdRequirer,
    controller_ready,
)
```

## Direction

The `slurmrestd` interface implements a provider/requirer pattern.
The Provider is the `slurmrestd` application that receives controller data from `slurmctld`.

> The Requirer is the `slurmctld` application that provides authentication secrets and Slurm configuration to `slurmrestd`.

Reviewer: ```suggestion
The `slurmrestd` interface implements the provider/requirer integration pattern. The provider-side application, `slurmrestd`, provides the integration endpoint that the requirer-side application, `slurmctld`, needs. `slurmctld` then supplies authentication secrets and Slurm configuration to `slurmrestd` through the application databag associated with the integration instance between `slurmrestd` and `slurmctld`. `slurmrestd` consumes controller data to configure the REST API daemon.
```

```mermaid
flowchart TD

>     Requirer -- auth_secret_id, slurmconfig --> Provider

Reviewer: ```suggestion
flowchart TD
    B("slurmrestd\n(provider)") -- endpoint --> A("slurmctld\n(requirer)")
    A("slurmctld\n(requirer)") -- auth_secret_id, slurmconfig --> B("slurmrestd\n(provider)")

Author: Same as my earlier comment. I'm removing the `endpoint` field from the Mermaid chart since it's not exchanged through the integration databag, but instead something that defined on the application.

```

## Behavior

The `slurmctld` requirer provides controller data (authentication secret and Slurm configuration) to the
`slurmrestd` provider. The `slurmrestd` provider does not send data back; it only consumes controller

> data to configure the REST API daemon.

Reviewer: Can be merged with Direction section above; suggestion to should cover most of it.

Author: Hmm... these sections should stay separate imo because they address two separate topics. The "Direction" section describes the data that moves between the provider/requirer whereas the "Behavior" section describes the expected behavior of the provider/requirer applications. E.g. the provider is expected to wait until it receives the controllers list and authentication secrets from the requirer.

I can wordsmith this section to make the expected behavior more clear, but let me know if my justification for keeping these sections separate makes sense before I proceed with this work.

Reviewer: The difference between the two sections felt minimal and like they were repeating information but worded slightly differently. The distinction between the two would need to be stronger, but imo it would be complicated to describe direction without behavior and vice versa since the behavior includes "I provide X" and "I receive Y" which are going to be directional. 

I'm open to keeping them separate if the distinction between the two topics is clearer.

Author: Sure, I'll wordsmith. I think conciseness is key here. Direction should really be focused only on the movement of the data between the charms and not how the data should be handled by the applications.

## Usage

This package provides the integration interface implementation for the `slurmrestd` interface.
It enables `slurmrestd` (Slurm REST API daemon) applications to exchange data with

> `slurmctld` (Slurm central management daemon), including authentication secrets and the Slurm configuration.

Reviewer: You mention `slurmctld` here but refer to it as `slurmrestd` requirer below. The `sackd` and `slurmd` ones don't say `slurmctld` anymore.

Author: Good catch, I'll clean this up.

Author: Fixed.

The `slurmrestd` requirer provides controller data (authentication secret and Slurm configuration) to the
`slurmrestd` provider. The `slurmrestd` provider does not send data back; it only consumes controller

> data to configure the REST API daemon.

Reviewer: ```suggestion
The `slurmrestd` requirer supplies controller data (authentication secret and Slurm configuration) to the
`slurmrestd` provider. The `slurmrestd` provider does not send data back; it only consumes controller
data to configure the REST API daemon.
```
